com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderUniformCollection.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace UnityEditor.ShaderFoundry
+{
+    internal class ShaderUniformCollection
+    {
+        Dictionary<string, UniformDeclarationData> visitedUniforms = new Dictionary<string, UniformDeclarationData>();
+        List<UniformDeclarationData> shaderUniforms = new List<UniformDeclarationData>();
+
+        public bool ReadOnly { get; private set; }
+        public IEnumerable<UniformDeclarationData> Uniforms => shaderUniforms;
+
+        public void Add(UniformDeclarationData uniform)
+        {
+            if (ReadOnly)
+            {
+                Debug.LogError("ERROR: attempting to add uniform to readonly collection");
+                return;
+            }
+
+            if (visitedUniforms.TryGetValue(uniform.Name, out var existingUniform))
+            {
+                ValidateAreEquivalent(uniform, existingUniform);
+                return;
+            }
+            visitedUniforms.Add(uniform.Name, uniform);
+            shaderUniforms.Add(uniform);
+        }
+
+        public void AddRange(IEnumerable<UniformDeclarationData> uniforms)
+        {
+            foreach (var uniform in uniforms)
```

Comment:
Same as above, it may be worth it to add a ReadOnly check here as well.

Reply:
Same as above.